File: BaekJoon/Simulation/로봇청소기_14503.cpp
//
//  로봇청소기_14503.cpp
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/03/01.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: - 로봇청소기 14503
// MARK: Simulation C++ 문제풀이

#if 1
#include <iostream>
#include <vector>
using namespace std;

vector<vector<int>> G(51, vector<int>(51,0));
int N, M, Ans=0;
int dx[] = {-1,0,1,0};
int dy[] = {0,1,0,-1};

struct Robot {
    int posX = 0;
    int posY = 0;
    int dir = 0;
    
    Robot(int x, int y, int dir) {
        this->posX = x;
        this->posY = y;
        this->dir = dir;
    }
    
    void cleanPlace() {
        if(G[posX][posY]==0) {
            G[posX][posY] = 2;
            
//            printf("Clean %d,%d\n",posX,posY);
//            for(int i=0; i<N; i++) {
//                for(int j=0; j<M; j++) {
//                    printf("%d ",G[i][j]);
//                }
//                puts("");
//            }
            
            Ans++;
        }
    }
    
    bool checkPlace() {
        int nd = dir;
        for(int i=0; i<4; i++) {
            nd--;
            if(nd<0) nd=3;
            int nx = posX + dx[nd];
            int ny = posY + dy[nd];
            if(nx<0 || ny<0 || nx>=N || ny>=M || G[nx][ny]>=1) continue;
            movePlace(nx, ny, nd);
            return true;
        }
        return false;
    }
    
    void movePlace(int x, int y, int d) {
        this->posX = x;
        this->posY = y;
        this->dir = d;
    }
    
    bool moveRear() {
        int nx = posX - dx[dir];
        int ny = posY - dy[dir];
        if(nx<0 || ny<0 || nx>=N || ny>=M || G[nx][ny]==1) return false;
        movePlace(nx, ny, dir);
        return true;
    }
};

int main() {
    ios_base :: sync_with_stdio(0); cin.tie(0);
    cin>>N>>M;
    int X, Y, D; cin>>X>>Y>>D;
    Robot robot = Robot(X,Y,D);
    
    for(int i=0; i<N; i++) {
        for(int j=0; j<M; j++) {
            cin>>G[i][j];
        }
    }
    
    while(1) {
        robot.cleanPlace();
        if(!robot.checkPlace()) {
            if(!robot.moveRear()) {
                return !printf("%d\n",Ans);
            }
        }
    }
    
    return 0;
}
#endif

//    3 3
//    1 1 1
//    0 0 0
//    0 0 0
//    0 0 0

File: BackJoonPS/BackjoonAlgorithmPS/Implementation/로봇청소기_14503.swift
//
//  로봇청소기_14503.swift
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/03/02.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: - 로봇청소기 14503
// MARK: 구현 swift 문제풀이

/*
import Foundation

let arr = readLine()!.split(separator: " ").map { Int($0)! }
let N = arr[0], M = arr[1]
let arr2 = readLine()!.split(separator: " ").map { Int($0)! }
let X = arr2[0], Y = arr2[1], Dir = arr2[2]
var G = [[Int]](repeating: [Int](), count: N)
var Ans = 0

for i in 0..<N {
    G[i] = readLine()!.split(separator: " ").map { Int($0)! }
}

let dx = [-1,0,1,0]
let dy = [0,1,0,-1]

class Robot {
    var posX = 0
    var posY = 0
    var dir = 0
    
    init(_ x: Int, _ y: Int, _ d: Int) {
        self.posX = x
        self.posY = y
        self.dir = d
    }
    
    func clean() {
        if G[posX][posY] == 0 {
            G[posX][posY] = 2
            Ans += 1
        }
    }
    
    func checkPlace() -> Bool {
        var nowDir = dir
        for _ in 0..<4 {
            nowDir -= 1
            if nowDir < 0 { nowDir = 3 }
            let nx = posX + dx[nowDir]
            let ny = posY + dy[nowDir]
            if nx < 0 || ny < 0 || nx >= N || ny >= M { continue }
            if G[nx][ny] >= 1 { continue }
            move(nx,ny,nowDir)
            return true
        }
        return false
    }
    
    private func move(_ X: Int, _ Y: Int, _ dir: Int) {
        self.posX = X
        self.posY = Y
        self.dir = dir
    }
    
    func back() -> Bool {
        let nx = posX - dx[self.dir]
        let ny = posY - dy[self.dir]
        if nx < 0 || ny < 0 || nx >= N || ny >= M { return false }
        if G[nx][ny] == 1 { return false }
        move(nx,ny,self.dir)
        return true
    }
}

let robot = Robot(X,Y,Dir)

while true {
    robot.clean()
    if !robot.checkPlace() {
        if !robot.back() {
            print(Ans); exit(0)
        }
    }
}
*/
